python/tskit/tables.py

```diff
@@ -2470,3 +2470,172 @@ def drop_index(self):
         currently indexed this method has no effect.
         """
         self.ll_tables.drop_index()
+
+    def add_time(self, dt):
+        """
+        This function modifies a table collection IN PLACE by adding ``dt`` to
+        all times in the node table.
+        """
+        nodes_dict = self.nodes.asdict()
+        nodes_dict['time'] = nodes_dict['time'] + dt
+        self.nodes.set_columns(**nodes_dict)
+        migrations_dict = self.migrations.asdict()
+        migrations_dict['time'] = migrations_dict['time'] + dt
+        self.migrations.set_columns(**migrations_dict)
+
+    def graft(self, other, node_map, check_overlap=False):
+        """
+        Returns tables by grafting the table collection ``other`` onto this
+        table collection along the nodes specified in the dictionary ``node_map``,
+        which should be a dictionary whose keys are node IDs in ``other`` and values
+        are the equivalent node IDs in this set of tables.
+        Any populations of nodes that are newly added to this table collection
+        are considered new in the grafted tables, even if their IDs are the same.
+        Map is returned (TODO: say exactly what).
+        If ``check_overlap`` is True, then check whether the two tables, simplified
+        to the shared nodes, are equal. This check is optional because it is relatively
+        expensive. TODO: should it be optional?
+        """
+        # checking shift in time ago between self and other
+        dt = [self.nodes.time[j] - other.nodes[i].time for i, j in node_map.items()]
+        if len(set(dt)) > 1:
+            raise ValueError(
+                "Inconsistent time differences among the equivalent nodes.")
+        new_tables = self.copy()
+        tables2 = other.copy()
+        dT = int(dt[0])
```

Comment:
I don't get the time change bit and why it's necessary - I don't think it's been explained anywhere either? Perhaps we should state this as an input requirement - it's easy enough to do the time change outside, and then we don't have to verify it in here.

Reply:
The time change is necessary if your forward-in-time simulations after the split are run for a different number of generations. @petrelharp and I went back and forth with this and we decided that giving the `node_map` would be enough to figure out whether the times should change or not. But I can also see this being an input requirement.

Reply:
I thought that all times in a tree sequence essentially have to be in real (calendar) time, or a regular scaling of real time. As soon as you have overlapping generations, node time can't be measured in true generations anyway, because two separate edge paths from one node to another node could correspond to different numbers of true generations. I.e. it makes no sense to have different scalings in different parts of the tree at the same time.

Reply:
The problem arises with forward-in-time simulations, not thinking about overlapping generations specifically. Imagine you want to simulate the root of (Neanderthal, Modern humans), and then after the split, simulate independently modern humans history and neanderthal history. At the end you'll end up with two tree sequences that have some shared history (the root), but the all the times in the shared portion of these tree sequences will be different because in the tree sequences we record time ago.

Reply:
now that I'm thinking about the numerical precision issue, we probably should make this an input requirement...

Reply:
It's easy enough to say "these nodes should be the same up to a time shift", so I don't see the harm in allowing it?

Reply:
what are you worried about?

Reply:
I think it would be simpler if we did the time shift outside the method and said it's an input requirement. It's a one-liner, and the caller will know the context. If nothing else, it'll make this method easier to test.